### PR TITLE
Added kwargs align, model, model_params, normalize to tools.cluster (Issue #138)

### PR DESCRIPTION
**Pull request by rarredon**
Added kwargs to `tools.cluster` as requested in Issue #138. Found a minor bug in `hypertools.plot` (`warn` function was being called but not imported, changed to `warnings.warn`; I think I introduced this in my last PR). Added simple test cases for `test_cluster.py`.

### CODE REVIEW

File: hypertools/tools/cluster.py [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
-from sklearn.decomposition import PCA
+import warnings
 import numpy as np
 from .._shared.helpers import *
+from .reduce import reduce as reducer
+from .align import align as aligner
+from .normalize import normalize as normalizer
 
-def cluster(x, n_clusters=8, ndims=None):
+def cluster(x, n_clusters=8, ndims=None, model='IncrementalPCA',
```
Comment (andrewheusser) [inline hypertools/tools/cluster.py]: @jeremymanning - for the `plot` function (as well as `reduce` and `align`), we use the keyword `model` to specify the dimensionality reduction model.  for the `align` and `cluster` functions, this may be a little weird since you might expect `model` to refer to the alignment/clustering model respectively, instead of the dimensionality reduction model.  Do you have an opinion on how to handle this?

If we want to support different alignment/clustering models in the future, we could add a `align_model` and `cluster_model` flag.  We could also add a `reduce_model` flag that could do the same thing as `model`.

Comment (jeremymanning) [inline hypertools/tools/cluster.py]: good point.  how about renaming those flags after their respective functions?  so:
- rename `model` to `reduce`
- `align` stays the same, but supports `True`, `False` (which defaults to aligning with hyperalignment or doing nothing, respectively), or `SRM` (align using SRM) or `hyperalign` (align using hyperalignment)-- default: `False`
- `cluster` stays the same, but supports `True` (align using k-means), `False` (do nothing; default), and `'kmeans'` (align using k-means) -- and possibly other models in the future

example syntax:
`y = hyp.plot(x, reduce='TSNE', cluster='kmeans', align='SRM', ndims=5)`

now `y` is an SRM-aligned TSNE-reduced 5-dimensional version of `x`, and a 3D plot (of `y`) is displayed.

Comment (andrewheusser) [inline hypertools/tools/cluster.py]: sounds reasonable - we also need a way to specify model parameters. what about..

`y = hyp.plot(x, reduce='TSNE', reduce_params={'sparsity': 40}, cluster='kmeans', align='SRM', ndims=5)`

similarly, `cluster_params` and `align_params` would allow the user to pass params to the cluster and align models, respectively.

We probably want to keep `model` and `model_params` but if they are used, we could create a warning that will say that these arguments will be deprecated in a future version. or we could just continue to support them

Comment (jeremymanning) [inline hypertools/tools/cluster.py]: the `reduce_params` syntax (and `cluster_params` and `align_params`) looks good.  that also brings up a good point-- the example i posted won't work, since it doesn't specify the number of k-means clusters.  so it should actually be something like:
`y = hyp.plot(x, reduce='TSNE', cluster='kmeans', cluster_params{'n': 6}, align='SRM', ndims=5)`

Re: keeping `model` and `model_params`, that approach (outputting a warning and depreciating them in the future) sounds good to me.

Comment (rarredon) [inline hypertools/tools/cluster.py]: Sounds good. I will implement these suggestions.

Comment (rarredon) [inline hypertools/tools/cluster.py]: I decided not to include kwargs for `align_params`, `cluster`, and `cluster_params` since there is currently nothing implemented to handle these arguments. I changed `model` and `model_params` to `reducer` and `reducer_params`, respectively; I avoided using `reduce` since it is a Python keyword.

Comment (rarredon) [inline hypertools/tools/cluster.py]: If you want, I can include those keyword args and in the docstring write something like "To be implemented in a future version."

**Comment by rarredon** [comment]
My last commit should also resolve Issue #140 .

**Comment by andrewheusser** [comment]
@rarredon , thank you!  In general, this looks great.  @jeremymanning, what do you think about `reducer` as the kwarg for reduction model, since `reduce` is a special python word.  Another option would be `reduce_model`/`reduce_params` or `reduction_model`/`reduction_params`.

**Comment by jeremymanning** [comment]
Hi @rarredon and @andrewheusser.  I'm not a big fan of the name `reducer`-- I'm worried it'll get confusing.  What about just sticking with `model` and `model_params`?  That's a bit shorter than `reduce_model` or `reduction_model`, but has a similar feel...

**Comment by andrewheusser** [comment]
Sure, we could revert back.  I think the logic was that in the future, we may want to support other alignment models, clustering models and so, the keyword `model` used for the reduction model might be confusing.  But perhaps its not necessary to do at this point. So - stick with `model` `model_params`?

**Comment by jeremymanning** [comment]
I think we should do something like:
- reduction model: algorithm controlled via `model`, keyword parameters specified via `model_params` keyword (reasoning: few characters to type, and the `reduce` keyword is already taken)
- alignment: algorithm/params controlled via `align` and `align_params` (reasoning: matches `align` function name)
- clustering: algorithm/params controlled via `cluster` and `cluster_params` (reasoning: matches `cluster` function name).  note: currently only 'kmeans' is supported.

We could (in a future release) allow users to either pass in a string to specify the algorithm, or they could specify a function (with an API we'd need to spell out carefully...e.g. what are the inputs, what are the outputs, etc.).

**Comment by andrewheusser** [comment]
hey @rarredon - I'm happy to finish this up with @jeremymanning's requests above if you'd like. just let me know!  thanks for all your help so far

**Comment by rarredon** [comment]
@andrewheusser  I will be a little busy into next week or so. Feel free to take over. Thanks :smiley:

**Comment by jeremymanning** [comment]
checking in on this issue-- can we merge this pull request, or are there open issues that still need to be addressed?

**Comment by andrewheusser** [comment]
@jeremymanning - i need to make a few changes before this is ready to merge. will do it ASAP!

**Comment by jeremymanning** [comment]
we should also address this issue: https://github.com/ContextLab/hypertools/issues/144

**Comment by jeremymanning** [comment]
Hi @rarredon,

I apologize for the long delay in processing this.  We've been working on a major revamp of the API, and these changes have been wrapped up in that process.  I'm going to close this pull request-- we'll be releasing version 0.4.0 shortly, which addresses these issues.  Thank you for your contributions; our interactions have helped highlight some of the "problem spots" in the toolbox, and we really appreciate your help (and code)!
